File: csharp/bank-account/BankAccount.cs
public class BankAccount
{
    private readonly Lock sync = new();
    private bool closed = true;
    private decimal balance = 0;

    public void Open()
    {
        lock (sync)
        {
            EnsureClosed();

            closed = false;
            balance = 0m;
        }
    }

    public void Close()
    {
        lock (sync)
        {
            EnsureOpen();
            closed = true;
        }
    }

    public decimal Balance
    {
        get
        {
            lock (sync)
            {
                EnsureOpen();
                return balance;
            }
        }
    }

    public void Deposit(decimal change)
    {
        if (change < 0) throw new InvalidOperationException();

        lock (sync)
        {
            EnsureOpen();
            balance += change;
        }
    }

    public void Withdraw(decimal change)
    {
        if (change < 0) throw new InvalidOperationException();

        lock (sync)
        {
            EnsureOpen();
            if (change > balance) throw new InvalidOperationException();

            balance -= change;
        }
    }

    private void EnsureOpen()
    {
        if (closed) throw new InvalidOperationException();
    }

    private void EnsureClosed()
    {
        if (!closed) throw new InvalidOperationException();
    }
}

File: cpp/bank-account/bank_account.cpp
#include "bank_account.h"
#include <stdexcept>

namespace Bankaccount
{

    void Bankaccount::open()
    {
        std::lock_guard<std::mutex> lock(mutex);

        if (opened)
        {
            throw std::runtime_error("Account opened");
        }

        opened = true;
        b = 0;
    }

    void Bankaccount::close()
    {
        std::lock_guard<std::mutex> lock(mutex);

        if (!opened)
        {
            throw std::runtime_error("Account closed");
        }

        opened = false;
    }

    int Bankaccount::balance() const
    {
        std::lock_guard<std::mutex> lock(mutex);

        if (!opened)
        {
            throw std::runtime_error("Account closed");
        }

        return b;
    }

    void Bankaccount::deposit(int amount)
    {
        std::lock_guard<std::mutex> lock(mutex);

        if (!opened)
        {
            throw std::runtime_error("Account closed");
        }

        if (amount < 0)
        {
            throw std::runtime_error("Cannot deposit negative amount");
        }

        b += amount;
    }

    void Bankaccount::withdraw(int amount)
    {
        std::lock_guard<std::mutex> lock(mutex);

        if (!opened)
        {
            throw std::runtime_error("Account closed");
        }

        if (amount < 0)
        {
            throw std::runtime_error("Cannot withdraw negative amount");
        }

        if (amount > b)
        {
            throw std::runtime_error("Insufficient funds");
        }

        b -= amount;
    }
}
